WT-5925 rework format CONFIG file syntax for clarity

keithbostic opened the pull request.

keithbostic commented:
Here's where I left the CONFIG:
```
############################################
#  RUN PARAMETERS: V2
############################################
assert.commit_timestamp=0
assert.read_timestamp=0
backup=0
backup.incremental=off
btree.bitcnt=1
btree.compression=none
btree.dictionary=0
btree.huffman_key=0
btree.huffman_value=0
btree.internal_key_truncation=1
btree.internal_page_max=13
btree.key_gap=8
btree.key_max=116
btree.key_min=32
btree.leaf_page_max=15
btree.memory_page_max=6
btree.prefix_compression=1
btree.prefix_compression_min=4
btree.repeat_data_pct=17
btree.reverse=0
btree.split_pct=81
btree.value_max=2106
btree.value_min=3
cache=81
cache.minimum=0
checkpoint=on
checkpoint.log_size=186
checkpoint.wait=48
data_source=table
disk.checksum=uncompressed
disk.data_extend=0
disk.direct_io=0
disk.encryption=none
disk.firstfit=0
disk.mmap=1
disk.mmap_all=0
evict_max=1
file_type=row-store
logging=1
logging.archive=0
logging.compression=none
logging.file_max=483366
logging.prealloc=1
lsm.auto_throttle=1
lsm.bloom=1
lsm.bloom_bit_count=28
lsm.bloom_hash_count=9
lsm.bloom_oldest=0
lsm.chunk_size=3
lsm.merge_max=5
lsm.worker_threads=3
major_timeout=0
ops.alter=1
ops.compaction=0
ops.pct.delete=8
ops.pct.insert=0
ops.pct.modify=0
ops.pct.read=90
ops.pct.write=2
ops.prepare=0
ops.random_cursor=0
ops.rebalance=1
ops.salvage=1
ops.truncate=1
ops.verify=1
quiet=0
runs=0
runs.abort=0
runs.independent_thread_rng=1
runs.ops=1827164
runs.rows=684034
runs.threads=4
runs.timer=30
statistics=0
statistics.server=0
stress.aggressive_sweep=0
stress.checkpoint=0
stress.hs_sweep=0
stress.split_1=0
stress.split_2=0
stress.split_3=0
stress.split_4=0
stress.split_5=0
stress.split_6=0
stress.split_7=0
stress.split_8=0
transaction.frequency=91
transaction.isolation=snapshot
transaction.timestamps=0
wiredtiger.config=
wiredtiger.in_memory=0
wiredtiger.leak_memory=0
############################################
```

keithbostic commented:
> I'd still like data_source and file_type to be together. Not merged but just visually together. What about type.data_source and type.file just to get the lines next to each other.

I guess I could do `runs.file_type` and `runs.data_source`, if you feel strongly?

sueloverso commented:
I'm not going to lose sleep over it, but `runs.<name>` works for me and would make me happy.

keithbostic commented:
> I'm not going to lose sleep over it, but `runs.<name>` works for me and would make me happy.

kk, reworked as suggested, I've updated the final version above for review.

agorrod commented on test/format/config_compat.c:
```diff
@@ -0,0 +1,92 @@
+ *     Convert old names to new ones.
+ */
+void
+config_compat(const char **origp)
```
It would be useful to be able to output old style configuration files as well? Maybe with a command line option? That will give us a way to run an older version of test/format against a failing config to see when a failure started.

keithbostic replied:
> It would be useful to be able to output old style configuration files as well? Maybe with a command line option? That will give us a way to run an older version of test/format against a failing config to see when a failure started.

That would be hard to do. Do we do that often enough to do the work?

Alternatively, maybe we could backport format to previous releases? (I suspect that's likely less work.)

agorrod replied:
It's not really releases, so I don't think a backport will help. For example we are using test/format as a reproducer with git bisect for WT-5826. That's looking at older commits, not released versions.

It seems like you've got a mapping from old->new, could you just generate the inverse? It doesn't need to be maintained indefinitely, so even a shell script that does the inverse mapping would be fine with me.

sueloverso replied:
A script would definitely be the only way forward. We don't want to have to maintain anything or add some kind of conversion backward for new config items we add. This compat function is static as of the known configurations at this point in time.

keithbostic replied:
Sure, pushed in a3d1cbc81.

agorrod commented:
Actually - could you update CONFIG.stress and CONFIG.endian to the new format before merging?

keithbostic commented:
> Actually - could you update CONFIG.stress and CONFIG.endian to the new format before merging?

Sure, pushed in 32c91ef07.

keithbostic commented:
@agorrod 
> Actually - could you update CONFIG.stress and CONFIG.endian to the new format before merging?

I updated the rest of the scripts in the directory as well.

agorrod commented:
You have been unlucky with testing - but this is looking good to me.

lukech commented:
`data_srouce` is usually the 1st thing I check in a format CONFIG. If it's not "lsm", is it easy not to dump those LSM related settings?

lukech commented on test/format/config_compat.sed:
```diff
@@ -0,0 +1,89 @@
+s/^assert.commit_timestamp=/assert_commit_timestamp=/
```
To make this script Evergreen compatible (which has prepended timestamp strings in the CONFIG dump), should we remove `^`?

keithbostic replied:
@lukech 
> To make this script Evergreen compatible (which has prepended timestamp strings in the CONFIG dump), should we remove ^?

The anchoring`^` is necessary, some configuration strings are subsets of other configuration strings, people will have to strip the timestamp strings before running the sed script, if it becomes necessary to use those CONFIG files with previous releases.

keithbostic commented:
@lukech 
> `data_source` is usually the 1st thing I check in a format CONFIG. If it's not "lsm", is it easy not to dump those LSM related settings? Same for column store `file_type` and bloom related settings.

Not dumping the strings is easy to do, but it's potentially tricky because you have to retain enough strings so that if the CONFIG file is subsequently loaded, the right things happen. I think we could do that, but it wasn't really the point of this work.

keithbostic commented:
@sueloverso, @agorrod, @lukech: maybe this isn't such a good idea. Luke & Alex have both raised issues with backward compatibility, maybe this isn't as obviously better as I hoped.

We're busy now, let's just kill this and live with what we have.

keithbostic commented:
We're not pulling this in, there's not enough of an advantage over what we have now.

agorrod commented:
I've re-opened the PR. I like this change, and hate digging through format configuration files at the moment. If there is pain we can spread the work of sorting it out over time.

agorrod commented:
ping @keithbostic

keithbostic commented:
@agorrod 
> ping

Do you want this merged? (To be clear, I'm OK just dropping it.)

agorrod commented:
Yes please, probably on Monday in case there are any surprises (i.e no rush, I just pinged you to make sure you saw that I reopened it)

keithbostic commented:
@lukech, merging WT-5925 (the format changes). I'll be watching for failures, but just ping me or revert if it goes pear-shaped.
